HW4_Voskoboinikov/ultimate_protein_tools.py

```diff
@@ -0,0 +1,691 @@
+    'ucc': 's',
+    'uca': 's',
+    'ucg': 's',
+    'uau': 'y',
+    'uac': 'y',
+    'uaa': '*',
+    'uag': '*',
+    'ugu': 'c',
+    'ugc': 'c',
+    'uga': '*',
+    'ugg': 'w',
+    'cuu': 'l',
+    'cuc': 'l',
+    'cua': 'l',
+    'cug': 'l',
+    'ccu': 'p',
+    'ccc': 'p',
+    'cca': 'p',
+    'ccg': 'p',
+    'cau': 'h',
+    'cac': 'h',
+    'caa': 'q',
+    'cag': 'q',
+    'cgu': 'r',
+    'cgc': 'r',
+    'cga': 'r',
+    'cgg': 'r',
+    'auu': 'i',
+    'auc': 'i',
+    'aua': 'i',
+    'aug': 'm',
+    'acu': 't',
+    'acc': 't',
+    'aca': 't',
+    'acg': 't',
+    'aau': 'n',
+    'aac': 'n',
+    'aaa': 'k',
+    'aag': 'k',
+    'agu': 's',
+    'agc': 's',
+    'aga': 'r',
+    'agg': 'r',
+    'guu': 'v',
+    'guc': 'v',
+    'gua': 'v',
+    'gug': 'v',
+    'gcu': 'a',
+    'gcc': 'a',
+    'gca': 'a',
+    'gcg': 'a',
+    'gau': 'd',
+    'gac': 'd',
+    'gaa': 'e',
+    'gag': 'e',
+    'ggu': 'g',
+    'ggc': 'g',
+    'gga': 'g',
+    'ggg': 'g'
+}
+
+
+def read_seq_from_fasta(path_to_seq: str,
+                        use_full_name: bool = False,
+                        **_) -> dict:
+    """
+    Reads sequences from fasta file and returns dictionary.
+
+    Arguments:
+    - path_to_seq (str): path to file
+
+    Return:
+    - dict: dict of sequences names as keys and sequences themselves as values {'seq_name': 'sequence',}
+    """
+
+    with open(path_to_seq) as f:
+        out_dct = {}
+        for line in f:
+            line = line.strip()
+            if line.startswith('>'):  # check for first line in seq
+                if use_full_name:  # check if user set full name in fasta
+                    name = line[1:]  # take whole fasta properties (e.g. if names not unique)
+                else:
+                    name = line[1:].split()[0]
+            else:
+                out_dct[name] = out_dct.get(name, '') + line  # get value from dict (return '' if empty) and append str
+    return out_dct
+
+
+def get_sites_lengths(sites: list) -> dict:
+    """
+    Takes sites list and calculates their lengths.
+
+    Arguments:
+    - sites (list): list of sites (str)
+
+    Return:
+    - dict: dict of sites length {'site': 'length',}
+    """
+
+    sites_length_dct = {}
+    for site in sites:
+        sites_length_dct[site] = len(site)
+    return sites_length_dct
+
+
+def invert_dct(dct: dict) -> dict:
+    """
+    Inverts a dict.
+
+    Arguments:
+    - dct (dict): dict to be inverted
+
+    Return:
+    - dict: inverted dict
+    """
+
+    inv_dct = {}
+    for k, v in dct.items():
+        inv_dct[v] = inv_dct.get(v, []) + [k]  # get value from dict (return [] if empty) and append key
+    return inv_dct
+
+
+def is_protein_valid(seq: str) -> bool:
+    """
+    Checks if protein is valid.
+
+    Arguments:
+    - seq (str): seq to be checked
+
+    Return:
+    - bool, the result of the check
+    """
+
+    if set(seq).issubset(RNA_AA_TABLE):
+        return True
+    return False
```

Comment:
В данном случае получаются небольшие проблемы с логикой. Вы проверяете булево выражение какое булево значение оно выдает, и если условие выполнилось, то возвращаете булево значение, а иначе проходите мимо него и возвращате другое булево значение. Тут можно было бы гораздно проще - сразу возвращать результат булевой операции :)

```suggestion
def is_protein_valid(seq: str) -> bool:
    """
    Checks if protein is valid.

    Arguments:
    - seq (str): seq to be checked

    Return:
    - bool, the result of the check
    """

    return set(seq).issubset(RNA_AA_TABLE):
```